spikeinterface/preprocessing/silence_periods.py

```diff
@@ -0,0 +1,132 @@
+            "Each element of 'list_periods' must be array-like"
+
+        for periods in list_periods:
+            if len(periods) > 0:
+                assert np.all(np.diff(np.array(periods), axis=1) > 0), "t_stops should be larger than t_starts"
+                assert all(e < s for (_, e), (s, _) in zip(periods, periods[1:])), "Intervals should not overlap"
+
+        assert mode in available_modes, f"mode {mode} is not an available mode: {available_modes}"
+
+        if mode in ['noise']:
+            noise_levels = get_noise_levels(recording, return_scaled=False, 
+                                    num_chunks_per_segment=num_chunks_per_segment,
+                                    chunk_size=chunk_size, concatenated=True, seed=seed)
+        else:
+            noise_levels = None
+
+        BasePreprocessor.__init__(self, recording)
+        for seg_index, parent_segment in enumerate(recording._recording_segments):
+            periods = list_periods[seg_index]
+            rec_segment = SilencedPeriodsRecordingSegment(parent_segment, periods, mode, noise_levels)
+            self.add_recording_segment(rec_segment)
+        
+        self._kwargs = dict(recording=recording.to_dict(), list_periods=list_periods,
+                            mode=mode, noise_levels=noise_levels)
+
+
+class SilencedPeriodsRecordingSegment(BasePreprocessorSegment):
+
+    def __init__(self, parent_recording_segment, periods, mode, noise_levels):
+
+        BasePreprocessorSegment.__init__(self, parent_recording_segment)
+
+        self.periods = np.asarray(periods, dtype='int64')
+        self.periods = np.sort(self.periods, axis=0)
+        self.mode = mode
+        self.noise_levels = noise_levels
+
+    def get_traces(self, start_frame, end_frame, channel_indices):
+
+        traces = self.parent_recording_segment.get_traces(start_frame, end_frame, channel_indices)
+        traces = traces.copy()
+        num_channels = traces.shape[1]
+
+        if start_frame is None:
+            start_frame = 0
+        if end_frame is None:
+            end_frame = self.get_num_samples()
+
+        if len(self.periods) > 0:
+            new_interval = np.array([start_frame, end_frame])
+            lower_index = np.searchsorted(self.periods[:,1], new_interval[0])
+            upper_index = np.searchsorted(self.periods[:,0], new_interval[1])
+
+            if upper_index > lower_index:
+
+                intersection = self.periods[lower_index:upper_index]
+
+                for i in intersection:
+
+                    onset = max(0, i[0] - start_frame)
+                    offset = min(i[1] - start_frame, end_frame)
+
+                    if self.mode == 'zeros':
+                        traces[onset:offset, :] = 0
+                    elif self.mode == 'noise':
+                        traces[onset:offset, :] = self.noise_levels[channel_indices] * np.random.randn(offset-onset, num_channels)
```

Comment:
I think we need a seed somewhere to be able to reproduce the same signals.
The seed could based a mixing aseed number + the segment index + period index.